physics/clm_lake.f90

```diff
@@ -0,0 +1,5737 @@
+      real(kind_phys) :: tksatu(1,nlevsoil)      ! thermal conductivity, saturated soil [W/m-K]
+      real(kind_phys) :: tkmg(1,nlevsoil)        ! thermal conductivity, soil minerals  [W/m-K]
+      real(kind_phys) :: tkdry(1,nlevsoil)       ! thermal conductivity, dry soil (W/m/Kelvin)
+      real(kind_phys) :: csol(1,nlevsoil)        ! heat capacity, soil solids (J/m**3/Kelvin)
+
+      integer :: lake_points, snow_points, ice_points
+      character*255 :: message
+      logical, parameter :: feedback_to_atmosphere = .true. ! FIXME: REMOVE
+
+      ! Functionality to print extra values at problematic points specified by user
+      logical :: was_unhappy,is_unhappy
+
+      ! Points come from this file
+      character(*), parameter :: unhappy_txt = "unhappy.txt"
+
+      ! Special values of the unhappy_count to indicate data is unavailable
+      integer, parameter :: HAVE_NOT_READ_UNHAPPY_POINTS_YET = -1
+      integer, parameter :: FAILED_TO_READ_UNHAPPY_POINTS = -2
+
+      ! These "save" variables are protected by an OMP CRITICAL to
+      ! ensure they're only initialized once.
+
+      ! Number of unhappy points
+      integer, save :: unhappy_count = HAVE_NOT_READ_UNHAPPY_POINTS_YET
+
+      ! The latitude and longitude of unhappy points.
+      real(kind_phys), allocatable, save :: unhappy_lat(:),unhappy_lon(:)
+
+      integer :: month,num1,num2,day_of_month
+      real(kind_phys) :: wght1,wght2,Tclim
+
+      errmsg = ' '
+      errflg = 0
+
+      dtime=dtp
+
+      if(LAKEDEBUG) then
+        ! Have we read the unhappy points?
+        ! The first "if" ensures we don't initiate an OMP CRITICAL unless we have to.
+         if(unhappy_count==HAVE_NOT_READ_UNHAPPY_POINTS_YET) then
+            !$OMP CRITICAL
+
+            ! Check unhappy_count again since it probably changed
+            ! during the setup of the omp critical, when another
+            ! thread read in the unhappy points.
+            if(unhappy_count==HAVE_NOT_READ_UNHAPPY_POINTS_YET) then
+               call read_unhappy_points
+               if(unhappy_count>0) then
+1308              format("Read ",I0,' points from unhappy point list file "',A,'"!')
+                  print 1308,unhappy_count,unhappy_txt
+8031              format('Read unhappy xlat_d=',F20.12,' xlon_d=',F20.12)
+                  do i=1,unhappy_count
+                     print 8031,unhappy_lat(i),unhappy_lon(i)
+                  enddo
+               endif
+            endif
+            !$OMP END CRITICAL
+         endif
+         ! At this point, at least one thread should have read in the unhappy points.
+         if(unhappy_count==FAILED_TO_READ_UNHAPPY_POINTS .and. kdt<2) then
+            write(0,'(A)') "Could not read unhappy points. Will not print unhappy point data."
+         endif
+      endif
+
+        ! Initialize any uninitialized lake points.
+        call lakeini(kdt=kdt, ISLTYP=ISLTYP, gt0=gt0, snowd=snowd, weasd=weasd,           &
```

Comment:
No. That routine is run only on one thread, and its purpose is to handle the rare cases where your initialization, on each timestep, requires the full grid. The lakeini does *not* require the full grid, so it belongs in the run step.

Reply:
My understanding is that this call is happening at the beginning of the _run phase, because it needs variables available on the physics timestep, which aren't present during initialization. No? 
(This kinda seems like the perfect example of code that should be in the timstep_init phase)

Reply:
The timestep_init is not threaded because it is intended for those rare cases when you need all data from the MPI rank to initialize once per timestep (like init has, but per timestep).

This routine does *not* need the full MPI rank worth of data, so it *can* be threaded. That means it belongs in the run function.

Reply:
I think we are talking part one another.
Just before calling lakeini there is a comment "Initialize any uninitialized ...". Then in [lakeini() there is another comment](https://github.com/ufs-community/ccpp-physics/blob/7c9a217d12d96e199e8d3a06e6a2600e0f993c50/physics/clm_lake.f90#L5377) stating that this routine is a completion of the initialization that needs information available on the physics timestep. This is a bit confusing and maybe the crux of my misunderstanding. It sounds like the code is updating data used at subsequent time-steps.

If you are updating static-data, then moving to the timestep_init phase makes sense, but if not then no. It just appeared that way from the comments.

Reply:
The timestep_init is not thread parallelized. Anything that can be thread parallelized belongs in run, otherwise you're just slowing the model down for no reason. The timestep_init was added solely to support the rare case when you need single-threaded access every timestep. It is not for static data; static data belongs in init.

Reply:
I understand the spirit of timestep_init. The comments aren't clear, which lead to my confusion.